### PR TITLE
#23 Add date component

### PR DESCRIPTION
**Pull request by burieberry**
- User can select date from calendar or by filling the input field
- User can navigate the calendar using arrow buttons or month/year dropdowns
- Clearing the input field resets calendar to current month + year view
- User inputted date gets formatted on enter
- TO DO: Styling, internationalization, setting year range

### CODE REVIEW

File: addon/components/date.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+const YEARS = Array(...Array(100)).map((_, i) => `${i + 1940}`);
+
+export default Component.extend({
+  layout,
+  classNames: ['cs-component-date'],
+  attributeBindings: ['dataTestName:data-test-cs-component-date'],
+  dataTestName: true,
+  months: MONTHS,
+  years: YEARS,
+
```
Comment (robustdj) [inline addon/components/date.js]: You should wrap these actions in the `actions` hook. For example:

```js
actions: {
  changeSelected(ev) {
    // ...
  },

  changeCenter(unit, calendar, val) {
    // ...
  }
}
```

Then when you call the action, be sure to wrap it in quotes i.e. `{{action "changeSelected"}}`

You can read more here: https://guides.emberjs.com/release/templates/actions/

File: addon/components/date.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  changeSelected(ev) {
+    let value = String(ev.target.value);
```
Comment (robustdj) [inline addon/components/date.js]: What's the purpose of using `String` here? Isn't `ev.target.value` already a string?

File: addon/components/date.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    value = moment(value, 'MM-DD-YYYY');
+
+    if (!value.toISOString()) {
+      // if no value, reset calendar view to current month and year
+      value = moment(moment.now());
```
Comment (robustdj) [inline addon/components/date.js]: To get a now moment object, you can just call `moment()`

```js
value = moment();
```

File: tests/acceptance/components-test.js [hunk 4]
```diff
@@ -103,4 +104,42 @@ module('Acceptance | components', function(hooks) {
     assert.dom('[data-test-cs-component-email-validation]').hasText('');
     assert.dom('[data-test-cs-component-email-validation].hidden').exists({ count: 2 });
   });
+
+  test('date component', async function (assert) {
+    let month = moment(moment.now()).format('MMMM');
+    let year = moment(moment.now()).format('YYYY');
```
Comment (robustdj) [inline tests/acceptance/components-test.js]: You can get a now moment object with `moment()`

**Review by robustdj — CHANGES_REQUESTED** [review]
Overall this looks good, just have a few suggestions

**Comment by robustdj** [comment]
Should we be a little more explicit with the names of our components? `Date` could be a component that just renders a date. Perhaps `DateSelector` would be more descriptive. What do you think?

File: addon/templates/components/date.hbs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      <button
+        class="ember-power-calendar-nav-control"
+        onclick={{action calendar.actions.moveCenter 12 "month"}}
+        data-test-nav-arrow="year-forwards">
+        »
+      </button>
+    </nav>
+    {{calendar.days}}
+  {{/calendar.content}}
+</PowerDatepicker>
```
Comment (robustdj) [inline addon/templates/components/date.hbs]: This is a nice display of component composition! Good work!

File: addon/components/date.js (hunk 1 above)
Comment (burieberry) [reply addon/components/date.js]: When should we use this type of syntax to call a function as opposed to the one without the actions wrapper? I've seen both styles, but not sure what the difference is.

Comment (robustdj) [reply addon/components/date.js]: I think whenever a function is triggered by any user action, we should use the `action` hook. For anything else like initial state setup, we should use regular functions. Oh and also ember-concurrency function generators don't use the action hook even though those can sometimes be triggered by user actions.

**Review by robustdj — APPROVED** [review]

**Pull request merged**